### PR TITLE
#141 Add Torque Calcs MVP (Unvalidated)

### LINKED ISSUE #141: [RTOS] - Torque Calcs
### Description

Change torque calcs to be state dependent, they should be mapped differently for each mode.

### Acceptance Criteria

Each mode has unique torque calcs that correspond to what we want from that mode

### Proposed Solution

Make it so that in pit the torque can only be so high, in performance we get all the beans, etc.

### PR DESCRIPTION
**Pull request by Peyton-McKee**
## Changes

Added Torque Calcs for different drive states. Added computation for translated ERPM to RPM and RPM to MPH for torque limiting. Converted C++ code to hopefully working c code from old cerb. Added Torque limiting inputs via paddles when in efficiency mode 

## To Do

regen still needs to be done if were gonna try that 

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #141  (issue #)

### CODE REVIEW

**Review by nwdepatie — CHANGES_REQUESTED** [review]
TODO:
- Add comments to torque calc so its easier to figure out what is going on
- Make DTI a struct you pass around to act as an "object", this implementation limits us to only ever having 1 motor controller
- make functions for button callbacks based on state

File: Core/Src/dti.c [hunk 1]
```diff
@@ -42,6 +44,8 @@ dti_t* dti_init()
 	dti_router_queue = osMessageQueueNew(CAN_QUEUE_SIZE, sizeof(can_msg_t), NULL);
 	assert(dti_router_queue);
 
+	dti = mc;
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply Core/Src/dti.c:47]: ah ok that makes more sense i was confused with what was going on.

File: Core/Src/torque.c [hunk 2]
```diff
@@ -26,11 +30,62 @@
...
+		}
+		else {
+			float torque_derating_factor = fabs(0.5 + ((-0.5/PIT_MAX_SPEED) * mph)); // Linearly derate torque from 0.5 to 0 as speed increases
+			newVal = accel * torque_derating_factor;
+		}
+
+		// The following code is a moving average filter
+		uint16_t ave = 0;
+		uint16_t temp[ACCUMULATOR_SIZE];
+		memcpy(torque_accumulator, temp, ACCUMULATOR_SIZE); // Copy the old values to the new array and shift them by one
```
Comment (nwdepatie) [inline Core/Src/torque.c:69]: THis memcpy moving average is kinda dope ngl

File: Core/Src/torque.c [hunk 3]
```diff
@@ -26,11 +30,62 @@
 #define MIN_COMMAND_FREQ  60					  /* Hz */
 #define MAX_COMMAND_DELAY 1000 / MIN_COMMAND_FREQ /* ms */
 
+static uint16_t torque_accumulator[ACCUMULATOR_SIZE];
```
Comment (nwdepatie) [inline Core/Src/torque.c:33]: I would recommend making this a static variable _inside_ the function for calcs. This limits the namespace of the variable to only inside that function AND it exists for the lifetime of the program

File: Core/Src/torque.c [hunk 4]
```diff
@@ -74,4 +145,23 @@ void vCalcTorque(void* pv_params)
 		dti_set_torque(torque);
 	}
 }
- 
\ No newline at end of file
+
+void increase_torque_limit()
```
Comment (nwdepatie) [inline Core/Src/torque.c:149]: nit: maybe move these up in the file, I like the task to be at the very bottom so it is easy to find

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Change accumulator to be static inside of a function, also we need to test so I won't approve until it is somehow tested

**Review by nwdepatie — APPROVED** [review]
looks solid

**Pull request merged**